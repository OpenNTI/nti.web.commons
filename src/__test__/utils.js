/*eslint-env jest*/
import renderer from 'react-test-renderer';

export const render = (exp, ...options) => renderer.create(exp, ...options);
export const verify = (exp, ...options) => {
	const r = render(exp, ...options);
	expect(r.toJSON()).toMatchSnapshot();
	return r;
};
