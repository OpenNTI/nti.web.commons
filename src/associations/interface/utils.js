import Group from './Group';

export function groupDestinations (destinations) {
	let currentWrapper = null;

	return destinations.reduce((acc, destination) => {
		if (destination instanceof Group) {
			currentWrapper = null;
			acc.push(destination);
		} else {
			if (!currentWrapper) {
				currentWrapper = new Group();
				acc.push(currentWrapper);
			}

			currentWrapper.push(destination);
		}

		return acc;
	}, []);
}

export function flattenGroups (groups) {
	return groups.reduce((acc, group) => {
		if (group instanceof Group) {
			acc = [...acc, ...flattenGroups(group.items)];
		} else {
			acc.push(group);
		}

		return acc;
	}, []);
}


export function filterGroups (groups, fn) {
	return groups.reduce((acc, group) => {
		const filteredGroup = group.filter(fn);

		if (!filteredGroup.isEmpty) {
			acc.push(filteredGroup);
		}

		return acc;
	}, []);
}


export function groupItemsByParent (items) {
	function getLabel (item) {
		return item.label || item.title;
	}

	const parents = items.reduce((acc, item) => {
		const parent = item.item.parent();
		const id = parent.getID();

		if (acc.map[id]) {
			acc.map[id].push(item);
		} else {
			acc.order.push(id);
			acc.map[id] = new Group (getLabel(parent), [item]);
		}

		return acc;
	}, {map: {}, order: []});

	return parents.order.map(x => parents.map[x]);
}
