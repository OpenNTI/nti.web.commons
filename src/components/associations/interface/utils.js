import Group from './Group';

export function groupDestinations (destinations) {
	let currentWrapper = null;

	return destinations.reduce((acc, destination) => {
		if (destination instanceof Group) {
			currentWrapper = null;
			acc.push(destination);
		} else {
			if (!currentWrapper) {
				currentWrapper = new Group();
				acc.push(currentWrapper);
			}

			currentWrapper.push(destination);
		}

		return acc;
	}, []);
}

export function mapActive (active) {
	return active.reduce((acc, item) => {
		if (item.NTIID != null) {
			acc[item.NTIID] = true;
		} else if (typeof item === 'string') {
			acc[item] = true;
		}

		return acc;
	}, {});
}


export function flattenGroups (groups) {
	return groups.reduce((acc, group) => {
		if (group instanceof Group) {
			acc = [...acc, ...flattenGroups(group.items)];
		} else {
			acc.push(group);
		}

		return acc;
	}, []);
}
