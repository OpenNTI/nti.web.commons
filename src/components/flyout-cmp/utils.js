import {
	DEFAULT_VERTICAL,
	DEFAULT_HORIZONTAL,
	DEFAULT_SIZING,

	VERTICAL,
	// HORIZONTAL,

	ALIGN_TOP,
	ALIGN_BOTTOM,
	ALIGN_LEFT,
	ALIGN_CENTER,
	ALIGN_RIGHT,

	MATCH_SIDE
} from './Constants';

/**
 * Functions to compute the different types of alignments
 * @type {Object}
 */
export const ALIGNMENT_POSITIONS = {
	//TODO: add horizontal positioning
	[VERTICAL]: {
		/**
		 * Align the bottom of the flyout to the top of the trigger.
		 * Use the bottom positioning to allow it to grow upwards.
		 *
		 * 		|  Flyout  |
		 * 		------------
		 * 		------------
		 * 		|  Trigger |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[ALIGN_TOP] ({top}, flyout, {height:viewHeight}) {
			return {
				top: null,
				bottom: viewHeight - top
			};
		},

		/**
		 * Align the top of the flyout to the bottom of the trigger.
		 * Use the top positioning to allow it to grow downwards.
		 *
		 * 		|  Trigger |
		 * 		------------
		 * 		------------
		 * 		|  Flyout  |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[ALIGN_BOTTOM] ({bottom}) {
			return {
				top: bottom,
				bottom: null
			};
		},

		/**
		 * Align the flyout to the top or bottom based on how much available space if there.
		 *
		 * If the height of the flyout will fit below the trigger, put it below.
		 * Else if the height of the flyout will fit above the trigger, put it there.
		 * Else put it to which side has the most space.
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[DEFAULT_VERTICAL] ({top, bottom}, {offsetHeight: flyoutHeight}, {height: viewHeight}) {
			const topSpace = top;
			const bottomSpace = viewHeight - bottom;

			const bottomAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_BOTTOM]({bottom});
			const topAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_TOP]({top}, void 0, {height: viewHeight});

			let position;

			if (bottomSpace >= flyoutHeight) {
				position = bottomAlignment;
			} else if (topSpace >= flyoutHeight) {
				position = topAlignment;
			} else {
				position = bottomSpace >= topSpace ? bottomAlignment : topAlignment;
			}

			return position;
		},

		/**
		 * Align the left of the flyout to the left of the trigger.
		 * Use the left position so it will grow right.
		 *
		 * 		|  Trigger |
		 * 		------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_LEFT] ({left}) {
			return {
				left: left,
				right: null
			};
		},

		/**
		 * Align the right of the flyout to the right of the trigger.
		 * Use the right position so it will grow left.
		 *
		 * 		    |  Trigger |
		 * 		    ------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_RIGHT] ({right}, flyout, {width: viewWidth}) {
			return {
				left: null,
				right: viewWidth - right
			};
		},

		/**
		 * Align the center of the flyout to the center of the trigger.
		 *
		 * 		  |  Trigger |
		 * 		  ------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * -OR-
		 *
		 * 		|  Trigger      |
		 * 		----------------
		 * 		  ------------
		 * 		  |  Flyout |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_CENTER] ({left, width:triggerWidth}, {offsetWidth: flyoutWidth}) {
			const triggerMid = Math.floor(triggerWidth / 2);
			const flyoutMid = Math.floor(flyoutWidth / 2);

			return {
				left: left + (triggerMid - flyoutMid),
				right: null
			};
		},


		/**
		 * Default to ALIGN_CENTER
		 *
		 * @type {Function}
		 * @return {Object} the horizontal positioning
		 */
		[DEFAULT_HORIZONTAL] (...args) {
			return ALIGNMENT_POSITIONS[VERTICAL][ALIGN_CENTER](...args);
		}
	}
};


export const ALIGNMENT_SIZINGS = {
	//TODO: add horizontal sizing
	[VERTICAL]: {
		[MATCH_SIDE] ({width}) {
			return {
				width: width
			};
		},
		[DEFAULT_SIZING] () {
			return {};
		}
	}
};


export function constrainAlignment (alignment = {}, {height: viewHeight, width: viewWidth}) {
	if (alignment.top != null) {
		alignment.maxHeight = viewHeight - alignment.top;
	} else if (alignment.bottom != null) {
		alignment.maxHeight = viewHeight - alignment.bottom;
	}

	if (alignment.left != null) {
		alignment.maxWidth = viewWidth - alignment.left;
	} else if (alignment.right != null) {
		alignment.maxWidth = viewWidth - alignment.right;
	}

	return alignment;
}
