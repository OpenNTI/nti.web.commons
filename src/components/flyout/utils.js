import {
	DEFAULT_VERTICAL,
	DEFAULT_HORIZONTAL,
	DEFAULT_SIZING,

	VERTICAL,
	// HORIZONTAL,

	ALIGN_TOP,
	ALIGN_BOTTOM,
	ALIGN_LEFT,
	ALIGN_CENTER,
	ALIGN_RIGHT,
	ALIGN_LEFT_OR_RIGHT,

	MATCH_SIDE
} from './Constants';

const ARROW_HEIGHT = 15;
const ARROW_OFFSET = 23;

const styleProps = ['top', 'bottom', 'left', 'right', 'width'];

/**
 * Functions to compute the different types of alignments
 * @type {Object}
 */
export const ALIGNMENT_POSITIONS = {
	//TODO: add horizontal positioning
	[VERTICAL]: {
		/**
		 * Align the bottom of the flyout to the top of the trigger.
		 * Use the bottom positioning to allow it to grow upwards.
		 *
		 * 		|  Flyout  |
		 * 		------------
		 * 		------------
		 * 		|  Trigger |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[ALIGN_TOP] ({top}, flyout, {height:viewHeight}) {
			return {
				top: null,
				bottom: viewHeight - top
			};
		},

		/**
		 * Align the top of the flyout to the bottom of the trigger.
		 * Use the top positioning to allow it to grow downwards.
		 *
		 * 		|  Trigger |
		 * 		------------
		 * 		------------
		 * 		|  Flyout  |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[ALIGN_BOTTOM] ({bottom}) {
			return {
				top: bottom,
				bottom: null
			};
		},

		/**
		 * Align the flyout to the top or bottom based on how much available space if there.
		 *
		 * If the height of the flyout will fit below the trigger, put it below.
		 * Else if the height of the flyout will fit above the trigger, put it there.
		 * Else put it to which side has the most space.
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the vertical positioning
		 */
		[DEFAULT_VERTICAL] ({top, bottom}, {offsetHeight: flyoutHeight}, {height: viewHeight}) {
			const topSpace = top;
			const bottomSpace = viewHeight - bottom;

			const bottomAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_BOTTOM](...arguments);
			const topAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_TOP](...arguments);

			let position;

			if (bottomSpace >= flyoutHeight) {
				position = bottomAlignment;
			} else if (topSpace >= flyoutHeight) {
				position = topAlignment;
			} else {
				position = bottomSpace >= topSpace ? bottomAlignment : topAlignment;
			}

			return position;
		},

		/**
		 * Align the left of the flyout to the left of the trigger.
		 * Use the left position so it will grow right.
		 *
		 * 		|  Trigger |
		 * 		------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_LEFT] ({left}) {
			return {
				left: left,
				right: null
			};
		},

		/**
		 * Align the right of the flyout to the right of the trigger.
		 * Use the right position so it will grow left.
		 *
		 * 		    |  Trigger |
		 * 		    ------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_RIGHT] ({right}, flyout, {width: viewWidth}) {
			return {
				left: null,
				right: viewWidth - right
			};
		},

		[ALIGN_LEFT_OR_RIGHT] ({left, right}, {offsetWidth: flyoutWidth}, {width: viewWidth}) {
			const leftSpace = left;
			const rightSpace = viewWidth - right;

			const leftAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_LEFT](...arguments);
			const rightAlignment = ALIGNMENT_POSITIONS[VERTICAL][ALIGN_RIGHT](...arguments);

			let position;

			if (rightSpace >= flyoutWidth) {
				position = leftAlignment;
			} else if (leftSpace >= flyoutWidth) {
				position = rightAlignment;
			} else {
				position = leftSpace >= rightSpace ? rightAlignment : leftAlignment;
			}

			return position;
		},

		/**
		 * Align the center of the flyout to the center of the trigger.
		 *
		 * 		  |  Trigger |
		 * 		  ------------
		 * 		----------------
		 * 		|  Flyout      |
		 *
		 * -OR-
		 *
		 * 		|  Trigger      |
		 * 		----------------
		 * 		  ------------
		 * 		  |  Flyout |
		 *
		 * @type {Function}
		 * @param {Object} triggerRect the rect for the trigger
		 * @param {Object} flyout the flyout dom node
		 * @param {Object} viewSize the size of the viewport
		 * @return {Object} the horizontal positioning
		 */
		[ALIGN_CENTER] ({left, width:triggerWidth}, {offsetWidth: flyoutWidth}) {
			const triggerMid = Math.floor(triggerWidth / 2);
			const flyoutMid = Math.floor(flyoutWidth / 2);

			return {
				left: left + (triggerMid - flyoutMid),
				right: null
			};
		},


		/**
		 * Default to ALIGN_CENTER
		 *
		 * @type {Function}
		 * @return {Object} the horizontal positioning
		 */
		[DEFAULT_HORIZONTAL] (...args) {
			return ALIGNMENT_POSITIONS[VERTICAL][ALIGN_CENTER](...args);
		}
	}
};


export const ALIGNMENT_SIZINGS = {
	//TODO: add horizontal sizing
	[VERTICAL]: {
		[MATCH_SIDE] ({width}) {
			return {
				width: width
			};
		},
		[DEFAULT_SIZING] () {
			return {};
		}
	}
};


export function constrainAlignment (alignment = {}, {height: viewHeight, width: viewWidth}) {
	const clone = {...alignment};

	if (clone.top != null) {
		clone.maxHeight = viewHeight - clone.top;
	} else if (clone.bottom != null) {
		clone.maxHeight = viewHeight - clone.bottom;
	}

	if (clone.left != null) {
		clone.maxWidth = viewWidth - clone.left;
	} else if (clone.right != null) {
		clone.maxWidth = viewWidth - clone.right;
	}

	return clone;
}


export function getOuterStylesForAlignment (alignment = {}, arrow, primaryAxis, alignToArrow) {
	const clone = {...alignment};

	if (primaryAxis === VERTICAL && arrow) {
		if (clone.top != null) {
			clone.top = clone.top + ARROW_HEIGHT;
		} else if (clone.bottom != null) {
			clone.bottom = clone.bottom + ARROW_HEIGHT;
		}

		if (alignToArrow) {
			if (clone.left != null) {
				clone.left = clone.left - ARROW_OFFSET;
			} else if (clone.right != null) {
				clone.right = clone.right - ARROW_OFFSET;
			}
		}
	}

	return styleProps.reduce((acc, prop) => {
		if (clone[prop] != null) {
			acc[prop] = `${clone[prop]}px`;
		}

		return acc;
	}, {});
}


export function getInnerStylesForAlignment (alignment, arrow, primaryAxis) {
	let {maxWidth, maxHeight} = alignment;

	if (primaryAxis === VERTICAL && arrow) {
		maxHeight = maxHeight - ARROW_HEIGHT;
	}

	//TODO: figure out what to do for horizontal alignment

	return {
		maxWidth: maxWidth ? `${maxWidth}px` : null,
		maxHeight: maxHeight ? `${maxHeight}px` : null
	};
}


export function getAlignmentClass (alignment, vAlign, hAlign) {
	//TODO: figure out what to do for horizontal alignment
	let vCls = '';
	let hCls = '';

	if (alignment.top != null) {
		vCls = 'bottom';
	} else if (alignment.bottom != null) {
		vCls = 'top';
	}

	if (!hAlign || hAlign === ALIGN_CENTER) {
		hCls = 'center';
	} else if (alignment.left != null) {
		hCls = 'left';
	} else if (alignment.right != null) {
		hCls = 'right';
	}

	return vCls && hCls ? `${vCls} ${hCls}` : (vCls || hCls);
}
