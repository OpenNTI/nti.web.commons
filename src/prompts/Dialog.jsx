import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Logger from '@nti/util-logger';
import uuid from 'uuid/v4';

import Manager from './ModalManager';

const logger = Logger.get('common:prompts:ManagedDialog');

/**
 * Managed Dialog Component Framework. Any child of this component will
 * be rendered into a modal portal.
 *
 * This sits atop the ModalManager which allows out-of-flow prompts
 * but this component locks the Modal to the lifecycle of ___this___ component. (By disallowing
 * dismissals without unmounting this instance)
 *
 * @class Dialog
 * @component
 */
export default class Dialog extends React.Component {

	static propTypes = {
		/** @ignore */
		children: PropTypes.any,
		/** @ignore */
		className: PropTypes.string,

		/** Set to false to ignore clicks/touches on the mask. */
		closeOnMaskClick: PropTypes.bool,

		/** Set to false to ignore the escape key */
		closeOnEscape: PropTypes.bool,

		/** Called to request dismissal of the modal. */
		onBeforeDismiss: PropTypes.func,

		/** Enables scroll memoization and restoration */
		restoreScroll: PropTypes.any,

		/**
		 * By default modals are centered on the screen. This will override that and align
		 * the modal to the top of the screen
		 */
		tall: PropTypes.bool,
	}

	mounted = true

	componentDidCatch (e) {
		/* istanbul ignore next */
		logger.error(e.stack || e.message || e);
	}


	componentDidMount () {
		this.uuid = uuid();
		this.rollbackCountStateKey = this.uuid + '-rollbackCount';

		global.addEventListener('popstate', this.onHistory);
		this.onHistory();
	}


	componentWillUnmount () {
		logger.debug('Un-mounting');
		this.mounted = false;
		const { rollbackCountStateKey: key } = this;
		const { history, document: { title } } = global;
		const { ...state } = history.state || {};

		// On Unmount remove our date and unregister our handler.
		global.removeEventListener('popstate', this.onHistory);

		delete state[key];

		history.replaceState(state, title);

		/* istanbul ignore else */
		if (this.m) {
			this.m.dismiss();
			delete this.m;
		}
	}


	onHistory = (e) => {
		const { rollbackCountStateKey: key } = this;
		const { history, document: { title } } = global;
		const {
			[key]: rollbackCount = null,
			...state
		} = history.state || {};

		this.rollbackCount = ((this.rollbackCount || 0) + 1); //local increment...

		// This looks unintuitive, but forward navigation does not have a 'state' value, so we have to track forward
		// motions internally. Events from "back/forward" will have the state of that entry (so we should take it as is)
		// So this code sets the state to itself, or our incremented value... then assigns that back to our internal
		// value.
		this.rollbackCount = state[key] = rollbackCount || this.rollbackCount;

		history.replaceState(state, title);

		logger.debug('History Entry changed: %o', state);
	}


	/* Call this to roll history back to before the dialog was opened */
	rollback () {
		const { rollbackCountStateKey: key } = this;
		const { history } = global;
		const { state } = history;
		const { [key]: rollbackCount } = state || {};

		logger.debug('Requested to roll-back history by: %o', rollbackCount);
		// If we have data in the history.state, use it instead...
		/* istanbul ignore else */
		if (rollbackCount > 0) {
			history.go(-rollbackCount);
		}
	}


	onBeforeDismiss = () => {
		const {onBeforeDismiss} = this.props;
		if (this.mounted) {
			logger.debug('Attempting to dismiss.');
			/* istanbul ignore else */
			if (onBeforeDismiss) {
				onBeforeDismiss({
					rollback: () => this.rollback()
				});
			}
			return false; //don't dismiss unless we unmount
		}
	}


	render () {
		const {mountPoint} = this.m || {};
		const {
			props: {
				children,
				className,
				closeOnMaskClick = true,
				closeOnEscape = true,
				restoreScroll,
				tall,
			}
		} = this;

		this.m = Manager.show(children,
			{
				className: cx('managed-modal', className),
				onBeforeDismiss: this.onBeforeDismiss,
				closeOnMaskClick,
				closeOnEscape,
				restoreScroll,
				mountPoint,
				tall,
				usePortal: true
			}
		);

		return this.m.portalRef;
	}
}
