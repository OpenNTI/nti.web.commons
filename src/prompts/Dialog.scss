.modal-mask.managed-modal {
	& > .icon-close { display: none; }
}
