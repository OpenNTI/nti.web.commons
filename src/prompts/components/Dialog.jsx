import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Logger from 'nti-util-logger';

import Manager from '../ModalManager';

const logger = Logger.get('common:prompts:ManagedDialog');

class DialogWrapper extends React.PureComponent {
	static propTypes = {
		children: PropTypes.node.isRequired,
	}


	render () {
		const {children, ...props} = this.props;
		const child = React.Children.only(children);

		if (child && typeof child.type === 'string') {
			delete props.onDismiss; //silence prop warnings on dom elements
		}

		return React.cloneElement(child, props);
	}
}


export default class ManagedDialog extends React.Component {

	static propTypes = {
		children: PropTypes.any,
		className: PropTypes.string,

		// Set to false to ignore clicks/touches on the mask.
		closeOnMaskClick: PropTypes.bool,

		//Use this to update your component's state to unmount the dialog.
		onBeforeDismiss: PropTypes.func,

		restoreScroll: PropTypes.any,

		tall: PropTypes.bool,
	}

	mounted = true

	componentDidCatch (e) {
		logger.error(e.stack || e.message || e);
	}


	componentWillUnmount () {
		logger.debug('Unmounting');
		this.mounted = false;
		if (this.m) {
			this.m.dismiss();
			delete this.m;
		}
	}


	onBeforeDismiss = () => {
		const {onBeforeDismiss} = this.props;
		if (this.mounted) {
			logger.debug('Attempting to dismiss.');
			if (onBeforeDismiss) {
				onBeforeDismiss();
			}
			return false; //don't dismiss unless we unmount
		}
	}


	render () {
		const {mountPoint} = this.m || {};
		const {
			props: {
				children,
				className,
				closeOnMaskClick = true,
				restoreScroll,
				tall,
			}
		} = this;

		this.m = Manager.show(
			<DialogWrapper {...{
				children
			}}/>,
			{
				className: cx('managed-modal', className),
				onBeforeDismiss: this.onBeforeDismiss,
				closeOnMaskClick,
				restoreScroll,
				mountPoint,
				tall,
				usePortal: true,
			}
		);

		return this.m.portalRef;
	}
}
