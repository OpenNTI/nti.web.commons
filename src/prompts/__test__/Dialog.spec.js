/* eslint-env jest */
jest.mock('react-dom', () => require('../../__mocks__/react-dom.disabled'));
jest.mock('@nti/lib-dom', () => Object.assign(
	require.requireActual('@nti/lib-dom'), {
		getScrollPosition: () => ({
			top: 100,
			left: 0,
		})
	}));

import React from 'react';

import {verify} from '../../__test__/utils';
import Dialog from '../Dialog';
import Manager from '../ModalManager';

const createNodeMock = ({type}) => document.createElement(type);

function getEscapeEvent () {
	const event = new CustomEvent('keydown');
	event.keyCode = 27;
	return event;
}

function mockScrollElement () {
	Object.defineProperty(document, 'scrollingElement', {
		configurable: true,
		value: {
			scrollTo: jest.fn()
		}
	});
}

describe('Dialogs', () => {
	document.body.innerHTML = '<div id="content"></div>';
	const content = document.body.firstChild;
	Object.defineProperty(document, 'activeElement', {configurable: true, value: content});

	beforeEach(() => {
		Manager.active.splice(0); // prevent each test from corrupting the next.
	});

	afterEach(() => {
		delete document.activeElement;
		delete document.scrollingElement;
	});

	test ('Simple Mount & Unmount', () => {
		jest.useFakeTimers();
		const Hi = () => <div>Hi</div>;
		const cmp = verify(<Dialog><Hi/></Dialog>, {createNodeMock});

		//Update lifecycle
		cmp.update(<Dialog><div>Hi2</div></Dialog>);


		expect(Manager.active.length).toBe(1);
		cmp.unmount();

		jest.runAllTimers();

		expect(Manager.active.length).toBe(0);
		expect(cmp.getInstance()).toBe(null);
	});

	test ('RestoreScroll on Unmount', () => {
		mockScrollElement();
		jest.useFakeTimers();

		Object.defineProperty(document, 'activeElement', {
			configurable: true,
			value: null
		});

		const cmp = verify(<Dialog restoreScroll alignTop><div>Hi</div></Dialog>, {createNodeMock});
		jest.runAllTimers();

		cmp.unmount();

		jest.runAllTimers();

		expect(document.scrollingElement.scrollTo).toHaveBeenCalledWith(0, 100);
	});

	test ('RestoreScroll custom fn on Unmount', () => {
		mockScrollElement();
		jest.useFakeTimers();

		const fn = jest.fn(() => true);
		const cmp = verify(<Dialog restoreScroll={fn} alignTop><div>Hi</div></Dialog>, {createNodeMock});

		jest.runAllTimers();

		cmp.unmount();

		jest.runAllTimers();

		expect(fn).toHaveBeenCalled();
		expect(document.scrollingElement.scrollTo).toHaveBeenCalledWith(0, 100);
	});


	test ('Escape Key Behavior', () => {
		jest.useFakeTimers();
		const fnA = jest.fn();
		const fnB = jest.fn();
		const btm = verify(<Dialog onBeforeDismiss={() => (fnA(), btm.unmount())}><div>Hi</div></Dialog>, {createNodeMock});
		const top = verify(<Dialog closeOnEscape={false} onBeforeDismiss={() => (fnB(), top.unmount())}><div>Yo</div></Dialog>, {createNodeMock});

		expect(Manager.active.length).toBe(2);

		document.dispatchEvent(getEscapeEvent());
		jest.runAllTimers();

		expect(fnA).not.toHaveBeenCalled();
		expect(fnB).not.toHaveBeenCalled();
		expect(Manager.active.length).toBe(2);

		top.unmount();
		jest.runAllTimers();

		expect(fnA).not.toHaveBeenCalled();
		expect(fnB).not.toHaveBeenCalled();
		expect(Manager.active.length).toBe(1);

		document.dispatchEvent(getEscapeEvent());
		jest.runAllTimers();

		expect(fnA).toHaveBeenCalled();
		expect(Manager.active.length).toBe(0);
	});


	test ('Focus Behavior', () => {
		jest.useFakeTimers();
		Manager.onDocumentFocus();

		const focus = jest.fn();
		const NodeMockFactory = (i) => {
			const ref = document.createElement(i.type);
			Object.defineProperty(ref, 'focus', {value: focus});
			return ref;
		};

		const btm = verify(<Dialog><div>Hi</div></Dialog>, {createNodeMock: NodeMockFactory});
		const top = verify(<Dialog><div>Yo</div></Dialog>, {createNodeMock: NodeMockFactory});

		const [{mountPoint: div}] = Manager.active;

		Manager.onDocumentFocus({target: div, stopPropagation () {}});

		jest.runAllTimers();

		expect(focus).toHaveBeenCalled();

		top.unmount();
		btm.unmount();
	});


	test ('Rollbock History', () => {
		jest.useFakeTimers();

		const fn = jest.fn(x => {
			// excercise callback ... should not throw,
			jest.spyOn(global.history, 'go');
			x.rollback();

			// simulate no rollback required (empty state)
			Object.defineProperty(global.history, 'state', {value: null});
			//should no-op and not throw
			x.rollback();
		});

		const cmp = verify(<Dialog onBeforeDismiss={fn}><div>Yo</div></Dialog>, {createNodeMock});

		document.dispatchEvent(getEscapeEvent());
		jest.runAllTimers();

		expect(global.history.go).toHaveBeenCalled();
		expect(fn).toHaveBeenCalled();

		cmp.unmount();
	});

});
