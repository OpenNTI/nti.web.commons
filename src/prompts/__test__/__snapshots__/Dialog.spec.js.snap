// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogs Escape Key Behavior 1`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Hi
        </div>
      </dialog>
    </div>
  </div>
</div>
`;

exports[`Dialogs Escape Key Behavior 2`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Yo
        </div>
      </dialog>
    </div>
  </div>
</div>
`;

exports[`Dialogs RestoreScroll custom fn on Unmount 1`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Hi
        </div>
      </dialog>
    </div>
  </div>
</div>
`;

exports[`Dialogs RestoreScroll on Unmount 1`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Hi
        </div>
      </dialog>
    </div>
  </div>
</div>
`;

exports[`Dialogs Rollbock History 1`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Yo
        </div>
      </dialog>
    </div>
  </div>
</div>
`;

exports[`Dialogs Simple Mount & Unmount 1`] = `
<div
  data-target-tag-name="DIV"
  data-type="portal"
>
  <div>
    <div
      className="modal-mask managed-modal"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <i
        className="icon-close"
        onClick={[Function]}
      />
      <dialog
        aria-modal="true"
        className="modal-content"
        onClick={[Function]}
        open={true}
        role="dialog"
        tabIndex="-1"
      >
        <div>
          Hi
        </div>
      </dialog>
    </div>
  </div>
</div>
`;
